### ISSUE TITLE
unmount fails due to null node when dom.nextSibling is null

### ISSUE THREAD

**Issue by scktt** [MithrilJS/mithril.js#1888]
sometimes when unmounting a view, removeNodeFromDOM fails when dom.nextSibling is null as it does not check that node is not null, causing the mount of the next view to also fail.

my changes to resolve the issue
```diff
--- a/root/js/mithril-1.1.1.js
+++ b/root/js/mithril-1.1.1.js
@@ -799,8 +799,10 @@ var coreRenderer = function($window) {
                }
        }
        function removeNodeFromDOM(node) {
-               var parent = node.parentNode
-               if (parent != null) parent.removeChild(node)
+               if(node != null) {
+                       var parent = node.parentNode;
+                       if (parent != null) parent.removeChild(node)
+               }
        }

```

**Comment by ArthurClemens** [comment]
I've come across that problem too.

**Comment by isiahmeadows** [comment]
Could you test it against Mithril's internal suite and file a PR?

**Comment by pygy** [comment]
I think the actual bug upstream, `node` should never be null in `removeNodeFromDOM`, by adding the proposed check we would be masking the actual bug...

Is `vnode.domSize` wrongly measured? Is it the bug in `continuation`? I'd love to have a repro for this.

**Comment by iainmcampbell** [comment]
Not sure if this is quite the same thing, but whatever it is I’ve managed to reproduce it: https://jsfiddle.net/qhprw0ku/7/. The exact error I’m getting is `Uncaught DOMException: Failed to execute 'removeChild' on 'Node': The node to be removed is no longer a child of this node. Perhaps it was moved in a 'blur' event handler?` I only get an error in Chrome 59, not Firefox 54 or Safari 10.

Call stack is `removeNodeFromDOM` &rarr; `continuation` &rarr; `removeNode` and so on.

To reproduce: create an `<input>`, add `onfocus` and `onblur` handlers, that add/remove a `keydown` event, that triggers a route change.

**Comment by isiahmeadows** [comment]
That is *super* odd, because the router is a very thin layer around `mount`, which simply throttles its calls to `render`.

**Comment by pygy** [comment]
@iainmcampbell I think it's another bug... Somehow redraw() ends up being called twice in the same tick, not sure why. [Adding an async layer](https://flems.io/#0=N4IgtglgJlA2CmIBcA2FA6FBOANCAzvAgMYAu8Uy4EpAFgE4SwACAjOu+wFb4h4BmTeLyQBtUADsAhmERIQ6HnxDEA9hPIaqANyn0ABGrAAHdfA2t9AXn3AAOhLul1ECTST7+AVwlkI6gAptCVUoeABKewl9fSddA3xSKXJrfToIfAcYmKzsxOT4dHV+VWIvfFTi0vLcmPzyIokAI1gvA0rm1vpa2Ojs719Sf2j1FraAyJ6YqGrZDXR6eDBVbXgAUVWNABkMzXh6AIByAGt4AE8ZgHcJQ5x9U4vVa-CegF8enoG-dX0qsvwJlFstkZmU5qR0FIYBtzKQdolzPsjg8rjc7iinhI7vwpLBCC8+jF3o5CZ4fN9ohjrgEIkDgfowAtVF4GoRSEcAPSkS6qQ4E4HEok4XK5bQQeCXDxfIaBYKhWki0jxfT1eCpdKZUmLUhtaKiD6kmJgI6uYws262KbZP7lDyqxolf7Cw3WzptDwqpINUZdZ30-SvcLOnoAXVyxOJDmVRlMElhACZUnSxRKpeSZRJAT1tbqGUcoBBtBbDtzefyAw5IxJlfRiKlQV5wegmqEzgslit4ABhWhMKAOBtNltQNuLZarHt91LS4Y02CTPpqCT4VQIdCwVQAcyOtYtRDuccu+jW9HoqgO4XQ+WIx3LOfo0VrkOMxlgZwCGruek3jdh+AJVaMmeLLwAEg6ws2rZ3IcHIWnSMGHB6MQxmYFh+voMGlohhiqCYqGkPGlYEiArwhngsCuMcIjiCAbSwFQtCkKQxj4EgHIcj4xjHJu6BGBykB0IwLCcBwyikGcxhyCASh4MYyTELQwjIKIoisHgni4vi+jKAACmekCEO2K6wKsEzoHQ5gBMoixQPQUiXHw8YAMw4AALAArDgTkAOysE58Zeb5-lkap6k2XZlwTJaEjKMCS7GYUG7boc4X2XyDjKE4TglAYQR6PoECOa5ODuQADIFrnxsVPmVa5IYhqRrxAA) solves the issue.

**Comment by pygy** [comment]
Also props to @porsager for making it so quick and easy to test hypotheses about bugs. Flems.io rules!

**Comment by isiahmeadows** [comment]
@pygy That probably goes along with our redraw work, then. @iainmcampbell Can you file a bug with that repro?

**Comment by isiahmeadows** [comment]
(I'm working on a simplified version that doesn't require `m.route`.)

**Comment by pygy** [comment]
Actually, it may be due to the fact that [m.route calls `run` directly](https://github.com/MithrilJS/mithril.js/blob/34c3430420d9db1493e2e7937519ea9b6b7bcddc/api/router.js#L27), and then the `onblur` event fires and triggers a sync `redraw`, since `redraw` is unaware of the `run()` call...

**Comment by pygy** [comment]
That is fixed in #1592 since `run()` is only called at mount time. After the first call `m.redraw()` takes over.

https://github.com/MithrilJS/mithril.js/pull/1592/files#diff-63de04924f7fd5b53de072769e289dfbR17

**Comment by isiahmeadows** [comment]
@scktt Could you provide your own repro to your issue, or is this basically another variant of @iainmcampbell's [repro](https://jsfiddle.net/qhprw0ku/7/)?

**Comment by iainmcampbell** [comment]
@pygy @isiahmeadows should I still open a new bug if this has already been addressed in #1592?

**Comment by isiahmeadows** [comment]
@iainmcampbell Probably so, since we didn't realize until now it was creating hard-to-debug exceptions for people in cases that it obviously *shouldn't* be. Most of the discussion up to this point has been just questions of API ergonomics and a few specific edge cases, not indirectly through a seemingly separate public API that *should* be handling that case.

**Comment by scktt** [comment]
@isiahmeadows will do, i can preproduce fairly quickly.

**Comment by pygy** [comment]
@scktt ping?

**Comment by scktt** [comment]
sorry @pygy for the delay. i have pulled the erroneous part of our app: https://jsfiddle.net/scktt/1wonq37e/5/

@isiahmeadows looking at the error it might be along the same lines but appears different as it doesnt appear to know the node had a child

**Comment by isiahmeadows** [comment]
@scktt Any chance you could cut it down some and remove the Bootstrap dependency?

- Bootstrap's JS could be installing crap that interferes with Mithril.
- That's a *lot* of code to sift through - two mount points, one using the router, and the Bootstrap boilerplate doesn't make it any easier.

Also, BTW, @iainmcampbell's issue was that of [unexpected, but technically correct behavior](https://github.com/MithrilJS/mithril.js/issues/1891#issuecomment-313746575) resulting in a recursive redraw call.

**Comment by scktt** [comment]
@isiahmeadows 

looks like the array inside the tabs view method is causing the problem. changing to

'''
view: function (vnode) {
  return "tabs"
}
'''

seems okay

https://jsfiddle.net/1wonq37e/6/

**Comment by pygy** [comment]
@scktt Great, thanks, I'll look at your sample later today

**Comment by pygy** [comment]
@scktt The problem stems from the fact that you're calling `m.mount()` from `oninit`. In its current implementation,`mount` calls redraw() whereas `m.route.set` doesn't and only renders the routed tree... The implicit `redraw` is unaware of the `route.set`-triggered `render()`, and that causes a nested `render()` call on the same root, a recipe for crashes and odd behaviours. That behaviour will be fixed in Mithril v2 (through #1592).

In the mean time, I'd advise you to do something [like this](https://flems.io/#0=N4IgzgpgNhDGAuEAmIBcIB0ArMIA0IAZgJYy6oDaoAdgIYC2EamAFvPVPiLAPbWL9mAHiTEAbgAJiSALwAdEPR4BXfgoB8QgPSix6udRHipshb361i1CACcN23epABfPDQZN02XAXMD4zGK0NhIQSqrwEjISSDywyoz8GADmEPAAojCJ8ABCAJ4AkkgAFArhaiAAlADcBkEhcHzwltYh0bHx2SlpmWEQ-PlFpdxNLbYKNQYGhKoIxHwStGC0AA7FhJUSwAAKNjz0xJAYNhBgPFBiEMWVGPAs-eubzlPU9RJQVgDWYABMURJbAwAgFiYgQADuqAkM2ocwWxTE1B4SAgm2AQOBAJO8GUNmoEnow1oCjwgPxwLkdxOhChCi0JIxmL4sBOtEQUPoxxUiAwH2on0ZElcEgUYBYPHBEh4d3GVQxz2oCrqwXeXzAAEZ-mTMaCIVCYXD8QikSi0YKsWlcfjCQpifhtZjgSxqbSQFpmgAjMAM8mOqWw1nsglc5Q8vkC33A4Wi8WSz3euXkhVK14qsUS-4AckzL055XgxTCKn4pPR5N14OuZcx2KtEnTkpkTYkmfDGszEgA-ATim31ZsOb21T9KvKDM5R9QXm8rCiAB7-av+qzEAuVJcAhtZvs5yNLVbFTknJA2WjgycA1xTeAV-WzeDzI2I5Goh2U2t4nuila0ai82getA-6AVAAC0KKELQyhQPAJIikQpAwHYiaXuO04qvGi4YnwK5rhu9axtuw67pi+5rEeyCnueIpJng163tC96PhIxovma5Lvpan4UOaX4gNBGB0GIgm0GIoHxnBPGRpiNogB8cHABIewwK6KwnJAFgPnwcGwFASxgK6tBzJcCjCrJdqlhIzoQDS8EAMSeuqcG2jYxC0KBfjKQmrqeqBTn2spEA+QBzkgEgbLufAPDJMkMAKMFHqmaSCgACoARI-mVJUdHScCsnyfaimBap6n9M0WlTi4pLmQpVkuvZno-KFwRuR5TRefF8G+U1AXnEFXUhfaCjhc04nRbFECdQonoKEKyUgGlHoSD1WW8QAutlNF+rJui3ABbX8GVkm8QCO3iA5AHqntHqgT+1i1SdNZ9Qld3QD6fqOi5rV6SBMBIB6eRTSAvn+TlH2Yrp+mGcZk0gI9V65TJwy7Xs4LveDeXI+IGC8GBYD0H5PXzYtGUTI9AIXh92WPWdYgXR6PzXbdv6w5Z5NYs9A0eq9UDoxjX3uT90B-QDQPdQo8NgxjtNcmj+Ds6dWPCbjoH44ToUkyt7OU36OvAhtgprWOiq5iGiCFn4Yw2PN9KFdedKdbOEBzlLdISWgEgSYqlQuGtBBtmgVD8TYnDoGw8ArAZWhaKoKyfMkOP7FoBxUqQAACV1XQAzMnq7OqQ3hcPAeQrJ4IA4L7zhAA)

There's an extra redraw scheduled on route change to update the `emount` mount point.

Alternatively, now that I think of it you could also `asap(()=>m.mount(emount, linksN)` in `oninit`, it will be less work on your end to refactor your app.

**Comment by scktt** [comment]
@pygy thanks for that, will do cheers
